src/methods/join.js

```diff
@@ -709,25 +717,26 @@ export class JoinImplementation {
     return result;
   }
 
-  // Folds two maps into a map and a list, based on an operator that joins map entries (e.g. bindings)
+  // Folds two maps into a map and a set, based on an operator that joins map entries (e.g. bindings)
   // but also spills some values to be dealt with later (e.g. objects)
   foldMaps<K, V, O>(
     acc: MapAndSet<K, V, O>,
     m1: Map<K, V>,
     m2: Map<K, V>,
     fold: (K, void | V, void | V) => [V, O]
   ): MapAndSet<K, V, O> {
+    invariant(acc.set !== undefined);
     m1.forEach((aVal1, key, map1) => {
       let aVal2 = m2.get(key);
       let [aVal3, bVal] = fold(key, aVal1, aVal2);
       acc.map.set(key, aVal3);
-      if (acc.set !== undefined && bVal !== undefined) acc.set.add(bVal);
+      if (bVal !== undefined) acc.set.add(bVal);
```

Comment:
The previous version of this made sense in as much as that it avoided building a set unless it was actually wanted. Do you now make use of the set for every call of foldMaps?

Reply:
I teetered between the two options, both of which are unsatisfying. With one you get a redundant Set, with the other you spill the void type outwards, all the way to the LeakedObjects type, get asymmetry in the Set, Map tuple (in the MapAndSet data structure, one can be undefined but the other can't). That invariant was the reason for the redundant Set in the previous version. I dislike both options, and chances are, every time somebody points out the infelicity of one, I will rush towards the other.

Reply:
Perhaps this calls for two separate methods?

Reply:
I think I have found a good middle ground, and fixed a typing bug that helps eliminate some unnecessary annotations and invariants.